File: docs/static/examples/src/fireworks/java/Particle.java
import org.joml.Vector2f;
import org.joml.Vector4f;

import static io.github.bernhardfritz.p8g.Sketch.*;

public class Particle {

    private Vector2f position;
    private Vector2f velocity;
    private Vector2f acceleration;
    private int age;
    private int lifespan;
    private float mass;
    private Vector4f color;

    public Particle(Vector2f position, float mass, Vector2f velocity, int lifespan, Vector4f color) {
        this.position = position;
        this.mass = mass;
        this.velocity = velocity;
        this.lifespan = lifespan;
        this.color = color;
        this.acceleration = new Vector2f();
        this.age = 0;
    }

    public void applyForce(Vector2f force) {
        Vector2f da = new Vector2f();
        force.mul(mass, da);
        acceleration.add(da);
    }

    public void update() {
        Vector2f deltaVelocity = new Vector2f();
        acceleration.mul(deltaTime / 19.f, deltaVelocity);
        velocity.add(deltaVelocity);
        Vector2f deltaPosition = new Vector2f();
        velocity.mul(deltaTime / 19.f, deltaPosition);
        position.add(deltaPosition);
        acceleration.set(0.f , 0.f);
        if (age < lifespan) {
            age = age + 1;
        }
    }

    public void draw() {
        if (lifespan == -1) {
            fill(color.get(0), color.get(1), color.get(2), color.get(3));
        } else {
            float pct = 1.f - age / (float) lifespan;
            fill(color.get(0), color.get(1), color.get(2), pct * 255.f);
        }
        ellipse(position.get(0), position.get(1), mass * 8.f, mass * 8.f);
    }

    public boolean isDead() {
        return age == lifespan;
    }

    public Vector2f getPosition() {
        return position;
    }

    public Vector2f getVelocity() {
        return velocity;
    }
}

File: docs/static/examples/src/fireworks/js/particle.js
import { deltaTime, ellipse, fill } from 'p8g.js';
import Victor from 'victor';

export default class Particle {
  constructor(position, mass, velocity, lifespan, color) {
    this._position = position;
    this._mass = mass;
    this._velocity = velocity;
    this._lifespan = lifespan;
    this._color = color;
    this._acceleration = new Victor(0, 0);
    this._age = 0;
  }

  applyForce(force) {
    const da = force.clone().multiply(new Victor(this._mass, this._mass));
    this._acceleration.add(da);
  }

  update() {
    const deltaVelocity = this._acceleration
      .clone()
      .multiply(new Victor(deltaTime / 19, deltaTime / 19));
    this.velocity.add(deltaVelocity);
    const deltaPosition = this.velocity
      .clone()
      .multiply(new Victor(deltaTime / 19, deltaTime / 19));
    this.position.add(deltaPosition);
    this._acceleration.x = 0;
    this._acceleration.y = 0;
    if (this._age < this._lifespan) {
      this._age = this._age + 1;
    }
  }

  draw() {
    if (this._lifespan === -1) {
      fill(this._color);
    } else {
      const pct = 1 - this._age / this._lifespan;
      fill(this._color[0], this._color[1], this._color[2], pct * 255);
    }
    ellipse(this._position.x, this._position.y, this._mass * 8, this._mass * 8);
  }

  isDead() {
    return this._age === this._lifespan;
  }

  get position() {
    return this._position;
  }

  get velocity() {
    return this._velocity;
  }

  set velocity(velocity) {
    this._velocity = velocity;
  }
}
